clang/lib/CodeGen/Targets/AMDGPU.cpp

```diff
@@ -115,7 +115,13 @@ void AMDGPUABIInfo::computeInfo(CGFunctionInfo &FI) const {
 
 Address AMDGPUABIInfo::EmitVAArg(CodeGenFunction &CGF, Address VAListAddr,
                                  QualType Ty) const {
-  llvm_unreachable("AMDGPU does not support varargs");
+  const bool IsIndirect = false;
+  const bool AllowHigherAlign = true;
+  // Would rather not naturally align values
+  // Splitting {char, short} into two separate arguments makes that difficult.
```

Comment:
Not sure what this comment is saying. 4 is the only correct alignment for anything stack related

Reply:
Currently amdgpu takes a `struct {char x; short y;} example;` and splats it across two registers. Doesn't seem totally good to me but I'm trying to avoid changing the non-variadic calling convention at the same time.

This means foo(example) and bar (char, short) look the same in IR, thus they need to have consistent representations as far as va_arg is concerned. The specific case that caused me trouble here was trying to pass a double with four byte alignment as opposed to eight (that's what the the naturally align values comment relates to).

I'll see if I can capture this in an amdgpu specific test case and fix up the comment

Reply:
I thought all structs under 8 bytes were packed into i32/i64

Reply:
Nope. At least, not what I got under testing. That sounds like a good idea though.

Reply:
^structs < 8 bytes are packed into integers, but structs larger than that are spread across multiple parameters in a discarding-padding sense. Better testing revealed that one byte aligning things does not work out robustly, have had to take a different strategy to make all the cases work. However that does mean everything is 4 byte aligned now.